paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java

```diff
@@ -434,9 +487,17 @@ private Snapshot changelogOrSnapshot(long snapshotId) {
         if (earliest == null || latest == null || snapshot(latest).watermark() == Long.MIN_VALUE) {
             return null;
         }
+
+        Snapshot earliestSnapShot =
+                tryGetEarliestSnapshotLaterThanOrEqualTo(earliest, latest, false);
+
+        if (earliestSnapShot == null) {
+            return null;
+        }
+
         Long earliestWatermark = null;
         // find the first snapshot with watermark
-        if ((earliestWatermark = snapshot(earliest).watermark()) == null) {
+        if ((earliestWatermark = earliestSnapShot.watermark()) == null) {
             while (earliest < latest) {
```

Comment:
I think there are some risks. Earliest should be re-assigned.

Maybe you should just optimize `earliestSnapshot`. And use this method in many places.

Reply:
Given that some earliestSnapshot methods use binary search while others use linear, I extracted and optimized two kinds of paths. Please take a look.